contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,168 @@
+    error ModuleRequiresRootAccess(address sender);
+    error MoudleNotEnabled(address module);
+    error ModuleEnabledOnlyForRootAccess(address module);
+    error ModuleAccessMismatch(address module, bool requiresRootAccess, bool providedValue);
+    error ActionExecutionFailed(address safe, bytes32 metaHash, uint256 index);
+    error RootAccessActionExecutionFailed(address safe, bytes32 metaHash);
+
+    constructor(address initalOwner) {
+        _transferOwnership(initalOwner);
+    }
+
+    modifier onlyEnabledModule(ISafe safe) {
+        if (!enabledComponents[address(safe)][msg.sender].enabled) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param transaction TODO
+     * @return data TODO
+     */
+    function executeTransaction(
+        ISafe safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bytes[] memory data) {
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        if (ISafeProtocolModule(msg.sender).requiresRootAccess()) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        if (enabledComponents[address(safe)][msg.sender].rootAddressGranted) {
+            revert ModuleEnabledOnlyForRootAccess(msg.sender);
+        }
+
+        data = new bytes[](transaction.actions.length);
+        for (uint256 i = 0; i < transaction.actions.length; ++i) {
+            SafeProtocolAction memory safeProtocolAction = transaction.actions[i];
+            (bool isActionSuccessful, bytes memory resultData) = safe.execTransactionFromModuleReturnData(
+                safeProtocolAction.to,
+                safeProtocolAction.value,
+                safeProtocolAction.data,
+                0
+            );
+
+            // Need to revisit the approach below. If some actions fail, the transaction stiil succeeds.
+            // With current approach, even if one action fails, `data` will be empty bytes even for successful
+            // actions.
+            if (!isActionSuccessful) {
+                revert ActionExecutionFailed(address(safe), transaction.metaHash, i);
+            } else {
+                emit ActionExecuted(address(safe), transaction.metaHash, i);
+                data[i] = resultData;
+            }
+        }
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param rootAccess TODO
+     * @return data TODO
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
+    ) external override onlyEnabledModule(safe) returns (bytes memory data) {
+        SafeProtocolAction memory safeProtocolAction = rootAccess.action;
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        // Re-confirm if this check if needed and correct.
+        if (!ISafeProtocolModule(msg.sender).requiresRootAccess()) {
```

Comment:
We are inconsistent on these checks. Some of them we do as modifiers (i.e. `onlyEnabledModule`) others we do inline (root check). Do we have some guidelines on when we use what?

Reply:
As of now, the checks that are repeated across multiple functions are done via function modifier. i.e. `noZeroOrSentinelModule`, `onlyEnabledModule`.

Reply:
makes sense